fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/RecordQuerySelectorPlan.java

```diff
@@ -0,0 +1,343 @@
+            return visitor.visitLeave(this);
+        }
+
+        private RecordCursorResult<QueryResult> calculateCursorResult(final RecordCursorResult<QueryResult> innerResult) {
+            final long startTime = System.nanoTime();
+            if (innerResult.hasNext()) {
+                SelectorContinuation continuation = new SelectorContinuation(selectedPlanIndex, innerResult.getContinuation().toBytes(), false);
+                logTimer(startTime);
+                return RecordCursorResult.withNextValue(innerResult.get(), continuation);
+            } else {
+                // Inner is done - return a no-next result with wrapped continuation and the same no-next reason and isEnd.
+                SelectorContinuation continuation = new SelectorContinuation(selectedPlanIndex, innerResult.getContinuation().toBytes(), innerResult.getContinuation().isEnd());
+                logTimer(startTime);
+                return RecordCursorResult.withoutNextValue(continuation, innerResult.getNoNextReason());
+            }
+        }
+
+        private void logTimer(final long startTime) {
+            if (timer != null) {
+                timer.record(FDBStoreTimer.Events.QUERY_SELECTOR, System.nanoTime() - startTime);
+            }
+        }
+    }
+
+    public interface PlanSelector extends PlanHashable {
+        int selectPlan(List<RecordQueryPlan> plans);
+    }
+
+    private static class SelectorContinuation implements RecordCursorContinuation {
+        private long selectedPlanIndex;
+        @Nullable
+        private ByteString innerContinuation = null;
+        private boolean isEnd;
+
+        public SelectorContinuation(byte[] rawBytes) {
+            try {
+                if (rawBytes != null) {
+                    RecordCursorProto.SelectorPlanContinuation continuation = RecordCursorProto.SelectorPlanContinuation.parseFrom(rawBytes);
+                    if (continuation.hasSelectedPlan()) {
+                        selectedPlanIndex = continuation.getSelectedPlan();
+                    }
+                    if (continuation.hasInnerContinuation()) {
+                        innerContinuation = continuation.getInnerContinuation();
+                    }
+                }
+            } catch (InvalidProtocolBufferException ex) {
+                throw new RecordCoreException("error parsing continuation", ex)
+                        .addLogInfo("raw_bytes", ByteArrayUtil2.loggable(rawBytes));
+            }
+        }
+
+        public SelectorContinuation(long selectedPlanIndex, @Nullable byte[] innerContinuation, final boolean isEnd) {
+            this.selectedPlanIndex = selectedPlanIndex;
+            this.isEnd = isEnd;
+            if (innerContinuation != null) {
+                this.innerContinuation = ByteString.copyFrom(innerContinuation);
+            } else {
+                this.innerContinuation = null;
+            }
+        }
+
+        @Nullable
+        @Override
+        public byte[] toBytes() {
```

Comment:
Is this the right continuation if `isEnd = true` (I thought that should return `null`)?

We're also not always consistent about this, but in some cases, we use `RecordCursorEndContinuation.END` for end continuations, and then the other continuation types always return `false` for `isEnd`, in which case this wouldn't need to handle the `isEnd = true` case.

Reply:
It seems clearer logic to delegate continuation state (including `isEnd()`) to the inner continuation, so just added `null` for the `isEnd` case in the selector continuation.
Done.